File: cases/taint/kotlin/dispatch-table-negative/DispatchTableNegative.kt
package dataflowbench

object DispatchTableNegative {
    fun dfb_source(): String { // DFB-SOURCE: dispatch-table-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: dispatch-table-sink

    fun leak(value: String) {
        dfb_sink(value)
    }

    fun drop(value: String) {
        dfb_sink("clean")
    }

    fun run() {
        val table = mapOf<String, (String) -> Unit>("leak" to ::leak, "drop" to ::drop) // DFB-WITNESS: dispatch-table-build
        val key = "drop"
        table.getValue(key)(dfb_source())
    }
}

File: cases/taint/cpp/dispatch-table-negative/dispatch_table_negative.cpp
#include <functional>
#include <map>
#include <string>

std::string dfb_source() { // DFB-SOURCE: dispatch-table-input
    return "tainted";
}

void dfb_sink(const std::string &value) {} // DFB-SINK: dispatch-table-sink

void leak(const std::string &value) {
    dfb_sink(value);
}

void drop(const std::string &value) {
    dfb_sink("clean");
}

void run() {
    std::map<std::string, std::function<void(const std::string &)>> table;
    table["leak"] = leak;
    table["drop"] = drop;
    std::string key = "drop";
    table[key](dfb_source());
}
